src/FastConv/FastConv.cpp

```diff
@@ -1,32 +1,192 @@
+
+// void FreqConvolution::multiplySpectra(float *spectrumC, const float *spectrumA, const float *spectrumB) {
+//     std::vector<float> realA(blockLength+1), imagA(blockLength+1);
+//     fft->splitRealImag(realA.data(), imagA.data(), spectrumA);
+
+//     std::vector<float> realB(blockLength+1), imagB(blockLength+1);
+//     fft->splitRealImag(realB.data(), imagB.data(), spectrumB);
+
+//     float scale = blockLength * 2;
+//     for (int i = 0; i < blockLength+1; i++) {
+//         float tmp = realA[i];
+//         realA[i] = (tmp * realB[i] - imagA[i] * imagB[i]) * scale;
+//         imagA[i] = (tmp * imagB[i] + imagA[i] * realB[i]) * scale;
+//     }
+//     fft->mergeRealImag(spectrumC, realA.data(), imagA.data());
+// }
+
+// Helper function: multiply two complex spectra, add result to output.
+static inline void addMultiplySpectra(float *output, const float *spectrumA, const float *spectrumB, int blockLength) {
+    // Spectrum layout: re(0),re(1),re(2),...,re(size/2),im(size/2-1),...,im(1)
+    output[0] += spectrumA[0] * spectrumB[0];
+    output[blockLength] += spectrumA[blockLength] * spectrumB[blockLength];
+    for (int i = 1; i < blockLength; i++) {
+        float realA = spectrumA[i];
+        float imagA = spectrumA[blockLength*2-i];
+        float realB = spectrumB[i];
+        float imagB = spectrumB[blockLength*2-i];
+        output[i] += realA * realB - imagA * imagB; // Real part
+        output[blockLength*2-i] += realA * imagB + imagA * realB; // Imaginary part
+    }
 }
 
-Error_t CFastConv::reset()
-{
+void FreqConvolution::process(float *output, const float *input, int length) {
+    int remainingInBlock;
+    // Fill and process as many complete blocks as we can.
+    while ((remainingInBlock = blockLength - indexInBlock) <= length) {
+        // Fill up the rest of the next input block.
+        std::copy_n(input, remainingInBlock, &inputBlocks[inputBlockIndex*blockLength*2 + indexInBlock]);
+        // Read out the rest of the last output block.
+        std::copy_n(&outputBlock[indexInBlock], remainingInBlock, output);
+
+        // We have another full input block for processing.
+        // Take the FFT in-place.
+        float *inputBlock = &inputBlocks[inputBlockIndex*blockLength*2];
+        std::fill(&inputBlock[blockLength], &inputBlock[blockLength*2], 0);
+        // NOTE: Unfortunately CFft adds some overhead here from copying buffers around,
+        // even though in this case we actually want the operation in-place.
+        // This is the best we can do without using LaszloFft directly.
+        fft->doFft(inputBlock, inputBlock);
+        // Save time-domain residue from last round & clear output block for accumulation.
+        std::copy_n(outputBlock.begin() + blockLength, blockLength, saved.begin());
+        std::fill(outputBlock.begin(), outputBlock.end(), 0);
+        // Sum results from circular convolutions (equivalent to linear convolutions due to zero-padding.)
```

Comment:
Using vector format for inputBlock and outputBlock seems convenient. Do you compare between whether using vector format instead of normal pointer declaration? I am not sure if using vector would take more time or not.

Reply:
Vectors are quite fast; if you're not resizing them frequently, then they're pretty much just heap-allocated arrays that know their own size (and remember to `delete` for you!).

In this code, we were previously using `memcpy` and `memset` directly on the array held by the vector, and I found that switching to `copy_in` and `fill` did not hurt performance. (There is some discussion & metrics about `memcpy` vs. `std::copy` [here](https://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance) if you're interested.)